incubator/nodejs/image/project/Dockerfile

```diff
@@ -8,14 +8,29 @@ RUN rm -rf /project/user-app/node_modules && mkdir -p /project/user-app/node_mod
 
 # Install user-app dependencies
 WORKDIR /project/user-app
-RUN npm install --production 
+RUN npm install --production
+
+# Run a build phase. Projects that need to execute build commands can customize the
+# 'build' script in their package.json. The build script should call 'npm install'.
+# If no build is required, just install production dependencies.
+RUN npm run build --if-present
+
+# Uninstall dev dependencies, leaving only production dependencies. Projects can
+# customize the 'prune' script in their package.json.
+# Ideally this would just be 'npm prune', but this command does not have pre/post
+# hooks. Instead, the user's 'prune' script can itself call 'npm prune' in addition
+# to any additional actions required.
+RUN npm run prune --production --if-present
```

Comment:
Again are we delegating specific tasks to developer now? will it work fine if I bring my own app that does not have these scripts defined?

Reply:
There is no change required for existing apps - the `--if-present` flag allows this command to succeed (do nothing).

If you bring an app that worked in the existing nodejs stack, it will continue to work in this version.  If you bring an app that requires a build step (such as running the `tsc` compiler), you would now be able to use it with this stack by doing something like this in your `package.json`:
```
    "build": "npm install && tsc",
    "prune": "npm prune"
```
This works because `npm run build` is run without the `--production` flag, and so calling `npm install` will install the `devDependencies` defined in the package.
When `npm run prune` is run, the `--production` flag sets environment which propagates to the `npm prune` command.  This removes any dependencies that would not be installed by `npm install --production`.